### PR TITLE
Implement the `Sum` trait for all `AffineCurve`

### PR DESCRIPTION
**Pull request by weikengchen**
## Description

George Gkitsas mentions that it would be convenient to have a `Sum` trait for `AffineCurve`, specifically SW curve's `GroupAffine`.

This PR adds the requirement that all `AffineCurve` must implement the `Sum` trait. 

It is useful to have this trait: a user who uses `+` or `+=` in SW's AffineCurve indeed would pay performance overhead --- there are unnecessary conversions between ProjectiveCurve and AffineCurve during each addition. A `Sum` trait implementation that avoids such unnecessary conversions can significantly boost the performance.

closes: #288 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

Feels skippable:
- [ ] Wrote unit tests
- [ ] Updated relevant documentation in the code

### CODE REVIEW

**Review by ValarDragon — APPROVED** [review]
utACK

File: ec/src/models/short_weierstrass_jacobian.rs [hunk 1]
```diff
@@ -278,6 +278,22 @@ impl<P: Parameters> Default for GroupAffine<P> {
...
+        iter.map(|x| x.into_projective())
+            .sum::<GroupProjective<P>>()
+            .into()
+    }
+}
+
+impl<'a, P: Parameters> core::iter::Sum<&'a Self> for GroupAffine<P> {
+    fn sum<I: Iterator<Item = &'a Self>>(iter: I) -> Self {
+        iter.map(|x| x.into_projective())
+            .sum::<GroupProjective<P>>()
```
Comment (Pratyush) [inline ec/src/models/short_weierstrass_jacobian.rs:292]: It would be faster to use mixed addition, no?

Comment (weikengchen) [reply ec/src/models/short_weierstrass_jacobian.rs:292]: Good point. Let me double-check. Yes, it is very likely that the conversions from affine to projective may be unnecessary as well.

Comment (weikengchen) [reply ec/src/models/short_weierstrass_jacobian.rs:292]: (though the latter is cheap)

Comment (weikengchen) [reply ec/src/models/short_weierstrass_jacobian.rs:292]: Done in #291

**Pull request merged**